Feat #2963: Implemented Query Variables Support

TomasSilv opened the pull request:
Resolves #2963

This PR adds support for inline variables in SQL queries, allowing users to define 
and use variables directly within the query text using special comment syntax. Variables can be 
declared using single-line comments like -- %var% = value or block comments such as /* VARS: ... */.

During parsing, these variables are extracted and removed from the original query,
 then substituted with the appropriate value based on their type. The substitution logic handles 
 quoting for strings, preserves numeric and null values, supports JSON objects, and formats lists 
 in a SQL-compatible way — all without introducing syntax errors.

The implementation includes robust error handling for malformed declarations and a
 comprehensive test suite that covers a wide range of use cases: valid and invalid inputs,
  repeated variables, optional substitutions, strings with spaces, floats, nulls, JSON, and lists.
  (We also added examples and information in editor.md)


https://github.com/user-attachments/assets/c14ac31b-c84d-4786-88b7-7fcdf88ce61e

not-night-but commented on apps/studio/src/common/utils.ts:
```diff
@@ -322,3 +322,128 @@ export function toRegexSafe(input: string) {
+    variables[name] = value.trim();
+  }
+
+  // Variables inside blocks: /* VARS: ... */
+  const blockPattern = /\/\*\s*VARS:(.*?)\*\//gs;
+  let blockMatch;
+  while ((blockMatch = blockPattern.exec(query)) !== null) {
+    const blockContent = blockMatch[1];
+
+    const varPattern = /%(.*?)%\s*=\s*(.*)/g;
+    let varMatch;
+    while ((varMatch = varPattern.exec(blockContent)) !== null) {
+      const [_, name, value] = varMatch;
+      if (!name || !value || value.trim() === '') {
+        throw new Error(`Malformed variable: "${varMatch[0]}"`);
+      }
+      variables[name] = value.trim();
+    }
+  }
+
```
Same issue as above. Not sure how we want it to work but:

```
/*
	VARS:

	%name%



	=



	'Beekeeper''s Database'
*/
```

This matches fine in the system, I'm just not sure if we want to allow it? Do you have any thoughts on this?

not-night-but commented on apps/studio/src/common/utils.ts:
```diff
@@ -322,3 +322,128 @@ export function toRegexSafe(input: string) {
+  let insideDoubleQuote = false;
+  let i = 0;
+
+  while (i < query.length) {
+    const char = query[i];
+    const nextChar = query[i + 1];
+
+    // Handle escaping of single quotes
+    if (char === "'" && nextChar === "'") {
+      result += "''";
+      i += 2;
+      continue;
+    }
+
+    // Toggle quote flags
+    if (!insideDoubleQuote && char === "'") {
+      insideSingleQuote = !insideSingleQuote;
+      result += char;
+      i++;
+      continue;
+    }
+
+    if (!insideSingleQuote && char === '"') {
+      insideDoubleQuote = !insideDoubleQuote;
+      result += char;
+      i++;
+      continue;
+    }
```
Okay so a couple things:

Not all dialects necessarily use single/double quotes for strings, if you look at our dialect datas  in `apps/studio/src/shared/lib/dialects/` you can see that each dialect may have its own way of escaping strings/quoting identifiers/etc. So I'm not sure we should be hardcoding this.

But regardless of that, how do other applications handle this? I believe the original issue mentions dbeaver. Does it allow for replacing within a string? Does it allow for a variable to be an identifier? eg:
```sql
-- %var% = name

SELECT %var% FROM connection;
```
Does that work in dbeaver? Is it something we should think about?

I want us to avoid accidentally malforming the query obviously, but we should look at what is supported in those other systems so we can attempt feature parity. Interested to hear your thoughts on this.

not-night-but commented on apps/studio/src/common/utils.ts:
```diff
@@ -322,3 +322,128 @@ export function toRegexSafe(input: string) {
+        // Not JSON
+      }
+
+      const isList = value.startsWith('(') && value.endsWith(')');
+      const isQuoted = /^'.*'$/.test(value);
```
Is any of this dialect specific? Can this be reliably used with postgres, mysql, bigquery, etc? If you need to make dialect specific changes for this, I would add the information you need to the dialect data for that dialect.

not-night-but reviewed (changes requested):
Okay this looks good ! Thanks so much for the contribution. I had a few questions, let me know your thoughts on them. 

Also, thank you so much for testing/documenting your feature! That helps out a bunch. Yall are legends!

TomasSilv commented:
Hi @not-night-but , thanks a lot for the detailed review and thoughtful suggestions!

Apologies for the delay — I’ve been a bit tied up recently, but I’m currently working on addressing your comments.

You're absolutely right about the regex pattern — that's something I’ll definitely need to improve to avoid multiline matches.
Regarding dialect-specific behavior, I’m not entirely sure yet how to approach that — right now it’s not fully adapted for all dialects, so that’s something I’ll need to explore further, especially considering how DBeaver handles similar cases.

I’ll push the updates as soon as I have everything tested and stable. Thanks again for your patience!
